Xamarin.Forms.Platform.Android/Material/MaterialButtonRenderer.cs

```diff
@@ -63,123 +67,51 @@ public MaterialButtonRenderer(Context context) : base(new ContextThemeWrapper(co
 			Tag = this;
 		}
 
-		public Color BackgroundColor => Element?.BackgroundColor != Color.Default ? Element.BackgroundColor : Color.Black;
-		public VisualElement Element => Button;
-		AView IVisualElementRenderer.View => this;
-		ViewGroup IVisualElementRenderer.ViewGroup => null;
-		VisualElementTracker IVisualElementRenderer.Tracker => _tracker;
+		protected MButton Control => this;
 
-		Button Button
+		protected Button Element
 		{
 			get => _button;
 			set
 			{
-				_button = value;
-				_platformElementConfiguration = null;
-			}
-		}
-
-		AView ITabStop.TabStop => this;
-
-		void IOnClickListener.OnClick(AView v) => ButtonElementManager.OnClick(Button, Button, v);
-
-		bool IOnTouchListener.OnTouch(AView v, MotionEvent e) => ButtonElementManager.OnTouch(Button, Button, v, e);
-
-		void IOnAttachStateChangeListener.OnViewAttachedToWindow(AView attachedView)
-		{
-			UpdateText();
-		}
-
-		void IOnAttachStateChangeListener.OnViewDetachedFromWindow(AView detachedView)
-		{
-		}
-
-		void IOnFocusChangeListener.OnFocusChange(AView v, bool hasFocus)
-		{
-			((IElementController)Button).SetValueFromRenderer(VisualElement.IsFocusedPropertyKey, hasFocus);
-		}
-
-		SizeRequest IVisualElementRenderer.GetDesiredSize(int widthConstraint, int heightConstraint)
-		{
-			UpdateText();
-
-			AView view = this;
+				if (_button == value)
+					return;
 
-			// with material something is removing the padding between it being set and 
-			// the measure call that's requested here
-			UpdatePadding();
-			view.Measure(widthConstraint, heightConstraint);
+				var oldElement = _button;
 
-			return new SizeRequest(new Size(MeasuredWidth, MeasuredHeight), MinimumSize());
-		}
-
-		void IVisualElementRenderer.SetElement(VisualElement element)
-		{
-			if (element == null)
-			{
-				throw new ArgumentNullException(nameof(element));
-			}
-
-			if (!(element is Button))
-			{
-				throw new ArgumentException($"{nameof(element)} must be of type {nameof(Button)}");
-			}
-
-			VisualElement oldElement = Button;
-			Button = (Button)element;
+				_button = value;
+				_platformElementConfiguration = null;
 
-			Performance.Start(out string reference);
-
-			if (oldElement != null)
-			{
-				oldElement.PropertyChanged -= OnElementPropertyChanged;
-			}
+				Performance.Start(out string reference);
 
+				if (oldElement != null)
+				{
+					oldElement.PropertyChanged -= OnElementPropertyChanged;
+				}
```

Comment:
Agreed these should all be inside SetElement

@mattleibow how much would you say this class duplicates the FastRenderer Button?

Reply:
@PureWeen I would say a fair amount. Probably good to do something like you did with the entry and have a generic base. One thing that may differ is the appearance bits - such as the colors and borders. But, that is mostly fine with a simple virtual function to replace the logic of the text/background colors.

Let me update with a generic base - should be ok since material button inherits from appcompat button.